### PR TITLE
Dcml annotation parser

### PR DESCRIPTION
**Pull request by manoskary**
### Import for DCML corpora tsv files

**DONE**
- functional import with the possibility to include only notes, or notes, measures and harmony.
- Added optional dependency in pandas
- Some updates on score classes such as Cadences, Roman numerals and Phrases.
- Updated the estimate symbolic duration function in `music.py`.
- Move globals from `partitura.utils.music` to separate script called `partitura.utils.globals` to avoid circular import errors. (This will be done in a following PR after this one is merged to not break other components).
- Included infer root note, bass note from Roman Numeral and Chord Symbol strings. 
- Included validity checks for Roman Numeral Symbols and Cadence types.
- Updates and fixes on note transposition

**TODOs** (For new PRs)
- Remove pandas dependency: 
 This will be planned for later using `numpy.genfromtxt`. It can be done but the process is much more complicated because it transforms  many strings to byte strings and it will take some major restructuring. For now the code raises a warning that pandas is required. For the future, we replace with `numpy.genfromtxt` and replace the warning. A relevant Issue will be opened.
- Notebook and github discussion with documentation about RomanNumerals and Intervals and such

### CODE REVIEW

File: partitura/directions.py [hunk 1]
```diff
@@ -51,27 +52,7 @@ def join_items(items):
...
-    (re.compile(r"(leggieramente|leggiermente|leggiero|legg\.?)"), "leggiero"),
-    (re.compile(r"(leggierissimo|(leggieriss\.?))"), "leggierissimo"),
-    (re.compile(r"(scherz\.?)"), "scherzando"),
-    (re.compile(r"(tenute|ten\.?)"), "tenuto"),
-    (re.compile(r"(allegretto)"), "allegro"),
-    (re.compile(r"(espress\.?)"), "espressivo"),
-    (re.compile(r"(ligato)"), "legato"),
-    (re.compile(r"(ligatissimo)"), "legatissimo"),
-    (re.compile(r"((rinforz|rinf|rfz|rf)\.?)"), "rinforzando"),
-]
```
Comment (manoskary) [inline partitura/directions.py:74]: Moved to globals.py

File: partitura/io/exportmusicxml.py [hunk 2]
```diff
@@ -834,6 +834,17 @@ def do_harmony(part, start, end):
...
+    # Does harmony annotation for cadences
+    # TODO: Merge with existing Roman Numeral and ChordSymbol annotations if they exist.
+    harmony = part.iter_all(score.Cadence, start, end)
+    for h in harmony:
+        harmony_e = etree.Element("harmony", print_frame="no")
+        function = etree.SubElement(harmony_e, "function")
+        function.text = "|" + h.text
+        kind_e = etree.SubElement(harmony_e, "kind", text="")
+        kind_e.text = "none"
+        result.append((h.start.t, None, harmony_e))
```
Comment (manoskary) [inline partitura/io/exportmusicxml.py:847]: Cadences are added to Roman Numeral labels or by themselves as `i | PAC` or `| PAC` as an exported harmony string. I also open to other suggestions.

File: partitura/io/importdcml.py
Comment (manoskary) [inline partitura/io/importdcml.py:1]: Main file for importing tsv files from dcml tabular format.
DCML includes usually three files, a note file, a bar/repeat/measure file, and an optional harmony file that includes cadences, phrases, and roman numerals. Partitura can only load the note content or optionally add the harmony and/or the bar files.

File: partitura/musicanalysis/key_identification.py [hunk 3]
```diff
@@ -11,80 +11,18 @@
...
-)
 
-# Kostka-Payne (from "Music and Probability" Table 6.1, pp. 86)
-key_prof_maj_kp = np.array(
-    [0.748, 0.060, 0.488, 0.082, 0.670, 0.460, 0.096, 0.715, 0.104, 0.366, 0.057, 0.400]
-)
-
-key_prof_min_kp = np.array(
-    [0.712, 0.048, 0.474, 0.618, 0.049, 0.460, 0.105, 0.747, 0.404, 0.067, 0.133, 0.330]
-)
```
Comment (manoskary) [inline partitura/musicanalysis/key_identification.py:87]: Moved to globals.py

File: partitura/musicanalysis/meter.py [hunk 4]
```diff
@@ -21,22 +21,12 @@
...
-CLUSTER_WIDTH = 1 / 12  # in seconds
-N_CLUSTERS = 100
-INIT_DURATION = 10  # in seconds
-TIMEOUT = 10  # in seconds
-TOLERANCE_POST = 0.4  # propotion of beat_interval
-TOLERANCE_PRE = 0.2  # proportion of beat_interval
-TOLERANCE_INNER = 1 / 12
-CORRECTION_FACTOR = 1 / 4  # higher => more correction (speed changes)
-MAX_AGENTS = 100  # delete low-scoring agents when there are more than MAX_AGENTS
-CHORD_SPREAD_TIME = 1 / 12  # for onset aggregation
```
Comment (manoskary) [inline partitura/musicanalysis/meter.py:39]: Moved to globals.py

File: partitura/utils/synth.py [hunk 5]
```diff
@@ -33,44 +34,6 @@
...
-    6: 7 / 5,
-    7: 3 / 2,
-    8: 8 / 5,
-    9: 5 / 3,
-    10: 9 / 5,
-    11: 15 / 8,
-    12: 2,
-}
-
-
```
Comment (manoskary) [inline partitura/utils/synth.py:73]: Moved to globals.py

File: partitura/utils/music.py [hunk 6]
```diff
@@ -40,297 +42,6 @@ class MIDITokenizer(object):
...
-
-SYM_COMPOSITE_DURS = [
-    ({"type": "quarter", "dots": 0}, {"type": "32nd", "dots": 0}),
-    ({"type": "quarter", "dots": 0}, {"type": "16nd", "dots": 0}),
-    ({"type": "half", "dots": 0}, {"type": "32nd", "dots": 0}),
-    ({"type": "half", "dots": 0}, {"type": "16th", "dots": 0}),
-    ({"type": "half", "dots": 0}, {"type": "eighth", "dots": 0}),
-]
-
-
```
Comment (manoskary) [inline partitura/utils/music.py:333]: Moved to globals

File: .github/workflows/partitura_unittests.yml [hunk 7]
```diff
@@ -28,7 +28,7 @@ jobs:
     - name: Install Optional dependencies
       run: |        
         pip install music21==8.3.0 Pillow==9.5.0 musescore==0.0.1      
-        pip install miditok==2.0.6 tokenizers==0.13.3 
+        pip install miditok==2.0.6 tokenizers==0.13.3 pandas==2.0.3
```
Comment (manoskary) [inline .github/workflows/partitura_unittests.yml:31]: added optional dependency for tests in load dcml.

File: partitura/io/importdcml.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+    alter[double_sharps] = 2
+    alter[double_flats] = -2
+    data["step"] = data["name"].apply(lambda x: x[0])
+    data["onset_div"] = onset_div
+    data["duration_div"] = duration_div
+    data["alter"] = alter
+    data["pitch"] = data["midi"]
+    grace_mask = ~data["gracenote"].isna().to_numpy() if "gracenote" in data.columns else np.zeros(len(data), dtype=bool)
+    data["id"] = np.arange(len(data))
+    # Rewrite Voices for correct export
```
Comment (fosfrancesco) [inline partitura/io/importdcml.py:47]: I would add a comment on this part, specifying what we do, i.e., taking the maximum voice number for the entire staff, and having the second staff starting from that number. 
The reason to be extra specific is that this is non-standard. For example, music notation software have voices starting from 1 on both staves. We will need to solve this with future partitura updates, but for now let's just make it very clear

File: partitura/io/importdcml.py [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+    for i, grace_el in enumerate(part.iter_all(spt.GraceNote)):
+        grace_idx = grace_note_idxs[i]
+        note = note_array[grace_idx]
+        # Find the next note in the same staff and voice
+        if not grace_mask[grace_idx+1]:
+            i = 1
+            next_note = note_array[grace_idx+i]
+            while note["staff"] != next_note["staff"] or note["voice"] != next_note["voice"]:
+                i += 1
+                next_note = note_array[grace_idx+i]
```
Comment (fosfrancesco) [inline partitura/io/importdcml.py:105]: add a ending condition for this search that will result in ignoring the grace note and print a warning

File: partitura/io/importdcml.py [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+    end_of_piece = (note_array["onset_div"]+note_array["duration_div"]).max()
+    end_divs = np.r_[start_divs[1:], end_of_piece]
+    for ts, start, end in zip(time_signatures, start_divs, end_divs):
+        part.add(spt.TimeSignature(beats=int(ts.split("/")[0]), beat_type=int(ts.split("/")[1])), start=start, end=end)
+
+    # Add default clefs for piano pieces (Naive)
+    part.add(spt.Clef(staff=1, sign="G", line=2, octave_change=0), start=0)
+    part.add(spt.Clef(staff=2, sign="F", line=4, octave_change=0), start=0)
+
+    # TODO: Find Ties
```
Comment (fosfrancesco) [inline partitura/io/importdcml.py:128]: is this TODO comment still valid?

File: partitura/io/importdcml.py [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+            assert note["voice"] == next_note["voice"], "Grace note and main note must be in the same voice"
+            assert note["onset_div"] == next_note[
+                "onset_div"], "Grace note and main note must have the same onset"
+            for note in part.iter_all(spt.Note, note["onset_div"], note["onset_div"]+1):
+                if note.id == "n-{}".format(next_note["id"]):
+                    grace_el.grace_next = note
+                    break
+
+    # Find time signatures
+    time_signatures_changes = data["timesig"][data["timesig"].shift(1) != data["timesig"]].index
```
Comment (fosfrancesco) [inline partitura/io/importdcml.py:116]: is the timesig in the main file? or in another? Just checking that everything is ok, ignore this comment if it's not relevant

File: partitura/score.py [hunk 12]
```diff
@@ -2773,20 +2810,228 @@ class RomanNumeral(TimedObject):
         See parameters
     """
 
-    def __init__(self, text):
+    def __init__(self, text, inversion=None, local_key=None, primary_degree=None, secondary_degree=None, quality=None):
+        super().__init__(text)
+        self.text = text
+        self.accepted_qualities = ('7', 'aug', 'aug6', 'aug7', 'dim', 'dim7', 'hdim7', 'maj', 'maj7', 'min', 'min7')
```
Comment (fosfrancesco) [inline partitura/score.py:2816]: Should this list be externalize in the global file or a similar place? Also what is aug7 ?

File: partitura/score.py [hunk 13]
```diff
@@ -2773,20 +2810,228 @@ class RomanNumeral(TimedObject):
...
+                return 2
+            elif inversion_state == 64:
+                self.has_seven = False
+                return 2
+            elif inversion_state == 6:
+                self.has_seven = False
+                return 1
+            elif inversion_state == 65:
+                self.has_seven = True
+                return 1
```
Comment (fosfrancesco) [inline partitura/score.py:2849]: add an else statement if the inversion is not part of the previous options

Comment (fosfrancesco) [reply partitura/score.py:2849]: this long if-elif can be made with a dictionary to be easier to read and maintain.
The key of the dictionary is the text, and the value is a tuple (has_seven,inversion).

File: partitura/score.py [hunk 14]
```diff
@@ -2826,6 +3071,26 @@ def validate(self):
             "down",
         ], f"Interval direction {self.direction} not found"
 
+    @property
+    def semitones(self):
+        return INTERVAL_TO_SEMITONES[self.quality + str(self.number)]
+
+    def change_quality(self, num):
```
Comment (fosfrancesco) [inline partitura/score.py:3078]: add documentation and comments to this function. It's not clear to me how it works

File: partitura/score.py [hunk 15]
```diff
@@ -5505,6 +5770,105 @@ def is_a_within_b(a, b, wholly=False):
     return contained
 
 
+def process_local_key(loc_k, glob_k, return_step_alter=False):
```
Comment (fosfrancesco) [inline partitura/score.py:5773]: rename loc_k to loc_k_text to avoid confusion

Comment (fosfrancesco) [reply partitura/score.py:5773]: and glob_k to glob_k_text

**Review by fosfrancesco — CHANGES_REQUESTED** [review]

**Comment by fosfrancesco** [comment]
The PR is good, I'm requesting some cosmetic changes for better code readability.

File: partitura/io/importdcml.py (hunk 9 above)
Comment (manoskary) [reply partitura/io/importdcml.py:105]: on it

File: partitura/io/importdcml.py (hunk 11 above)
Comment (manoskary) [reply partitura/io/importdcml.py:116]: Time signatures should be in the note file. With that line I am tracking time signature changes. With the shift time_sig[0] is set to NaN so the initial time signature is there.

File: partitura/score.py (hunk 12 above)
Comment (manoskary) [reply partitura/score.py:2816]: The tests are implemented within the class, so I would say it is better if it stays here, at least for the moment. Of course that is up to discusion

**Review by fosfrancesco — APPROVED** [review]

**Pull request merged**